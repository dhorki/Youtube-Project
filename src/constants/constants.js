export const styles = {
  colors: {
    light: {
      appIcon: '#ff0200',
      background: '#F9F9F9',
      barBackground: '#ffffff',
      primary: '#030303',
      secondary: '#606060',
      linkActive: '#ff0200',
      linkHoverBackground: '#ececec',
      inputBackground: '#ffffff',
      inputBorder: '#cccccc',
      searchButtonBackground: '#f8f8f8',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
    dark: {
      appIcon: '#ff0200',
      background: '#181818',
      barBackground: '#212121',
      primary: '#ffffff',
      secondary: '#aaaaaa',
      linkActive: '#ff0200',
      linkHoverBackground: '#303030',
      inputBackground: '#121212',
      inputBorder: '#303030',
      searchButtonBackground: '#323232',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
  },
  props: {
    transitionTime: '0.4s',
    sidebarWidth: '150px',
  },
};

export const screenSizes = {
  xsmall: '675px',
  small: '830px',
  mid: '1200px',
};

export const mainViewTypes = {
  searchGallery: 'searchGallery',
  favoritesGallery: 'favoritesGallery',
  videoDetails: 'videoDetails',
};

export const environmentActions = {
  setTheme: 'setTheme',
  setSidebarShow: 'setSidebarShow',
  setSearchQuery: 'setSearchQuery',
  setModalShow: 'setModalShow',
  setUser: 'setUser',
};

export const modalTypes = {
  login: 'login',
  register: 'register',
  resetPassword: 'reset-password',
  none: 'none',
};

export const resetSearchQueryAction = {
  type: environmentActions.setSearchQuery,
  payload: {
    q: '',
  },
};

export const hideModalAction = {
  type: environmentActions.setModalShow,
  payload: {
    modalShow: modalTypes.none,
  },
};
