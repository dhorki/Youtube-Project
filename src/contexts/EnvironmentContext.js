import { createContext } from 'react';

export const EnvironmentContext = createContext();
