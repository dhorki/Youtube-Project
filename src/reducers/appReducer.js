import { environmentActions } from '../constants/constants';

export const appReducer = (state, { type, payload }) => {
  switch (type) {
    case environmentActions.setTheme:
      return { ...state, theme: payload.theme };
    case environmentActions.setSidebarShow:
      return { ...state, sidebarShow: payload.sidebarShow };
    case environmentActions.setSearchQuery:
      return { ...state, q: payload.q };
    case environmentActions.setModalShow:
      return { ...state, modalShow: payload.modalShow };
    case environmentActions.setUser:
      return { ...state, user: payload.user };
    case environmentActions.setFavoritesList:
      const user = { ...state.user, favoritesList: payload.favoritesList };
      return { ...state, user: user };
    default:
      return state;
  }
};
