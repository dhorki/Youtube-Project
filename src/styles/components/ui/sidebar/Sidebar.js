import styled from 'styled-components';
import { styles } from '../../../../constants/constants';

export const StyledSidebar = styled.nav`
  background-color: ${({ theme }) => theme.barBackground};
  color: ${({ theme }) => theme.primary};
  top: 48px;
  left: ${({ toggle }) => (toggle ? '0px' : '-' + styles.props.sidebarWidth)};
  position: fixed;
  display: flex;
  flex-direction: column;
  z-index: 1;
  width: ${styles.props.sidebarWidth};
  height: 100vh;

  transition: background-color ${styles.props.transitionTime} ease;
  transition: left ${styles.props.transitionTime} ease;

  > .sidebar-link {
    text-decoration: none;
    color: ${({ theme }) => theme.primary};
    font-size: 15px;
    font-weight: bolder;
    display: flex;
    flex-direction: row;
    justify-content: left;
    align-items: center;
    width: 100%;
    min-width: 100px;
    height: 50px;

    &:hover {
      background-color: ${({ theme }) => theme.linkHoverBackground};
    }

    transition: background-color ${styles.props.transitionTime} ease;

    > .sidebar-link-icon {
      margin-left: 15px;
    }

    > span {
      margin-left: 10px;
      margin-right: 20px;
    }
  }

  > .sidebar-link-active {
    color: ${({ theme }) => theme.linkActive};
  }
`;
