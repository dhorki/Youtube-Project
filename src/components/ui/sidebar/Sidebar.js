import React, { useContext } from 'react';
import { NavLink } from 'react-router-dom';
import { AiFillHome } from 'react-icons/ai';
import { MdFavorite } from 'react-icons/md';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';
import { StyledSidebar } from '../../../styles/components/ui/sidebar/Sidebar';
import { resetSearch } from '../../../helpers/resetSearch';

export const Sidebar = () => {
  const { environment, dispatchEnv } = useContext(EnvironmentContext);
  const { theme, sidebarShow } = environment;

  return (
    <StyledSidebar theme={theme} toggle={sidebarShow}>
      <NavLink
        id="sidebar-home-link"
        className={(isActive) => `sidebar-link ${isActive ? 'sidebar-link-active' : ''}`}
        to="/"
        onClick={() => resetSearch(dispatchEnv)}
        exact
      >
        <AiFillHome className="sidebar-link-icon" />
        <span>Home</span>
      </NavLink>
      <NavLink
        id="sidebar-favorites-link"
        className={(isActive) => `sidebar-link ${isActive ? 'sidebar-link-active' : ''}`}
        to="/favorites"
        onClick={() => resetSearch(dispatchEnv)}
      >
        <MdFavorite className="sidebar-link-icon" />
        <span>Favorites</span>
      </NavLink>
    </StyledSidebar>
  );
};
