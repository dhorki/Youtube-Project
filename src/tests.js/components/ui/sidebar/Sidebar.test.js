import React, { useReducer } from 'react';
import { Route } from 'react-router-dom';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { styles } from '../../../../constants/constants';
import { EnvironmentContext } from '../../../../contexts/EnvironmentContext';
import { Sidebar } from '../../../../components/ui/sidebar/Sidebar';
import { renderHook, act } from '@testing-library/react-hooks';
import { appReducer } from '../../../../reducers/appReducer';
import { MemoryRouter } from 'react-router-dom';

describe('SideBar', () => {
  const environment = {
    theme: styles.colors.light,
    sidebarShow: false,
    q: '',
  };

  it('should render correctly with home route', () => {
    const mockLocation = {
      pathname: '/',
      search: '',
    };

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <Route path="/">
            <Sidebar />
          </Route>
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with favorites route', () => {
    const mockLocation = {
      pathname: '/favorites',
      search: '',
    };

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <Route path="/favorites">
            <Sidebar />
          </Route>
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should reset search query on home menu click', () => {
    environment.q = 'some search query';

    const mockLocation = {
      pathname: '/',
      search: '',
    };

    const { result } = renderHook(() => useReducer(appReducer, environment));
    const [prevEnv, dispatchEnv] = result.current;
    const { q: prevQ } = prevEnv;

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment: prevEnv, dispatchEnv }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <Sidebar />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );

    const link = wrapper.find('#sidebar-home-link').first();
    act(() => {
      link.simulate('click');
    });

    const [env] = result.current;
    const { q } = env;

    expect(q).not.toEqual(!prevQ);
    expect(q.length).toEqual(0);
  });

  it('should reset search query on favorites menu click', () => {
    environment.q = 'some search query';

    const mockLocation = {
      pathname: '/favorites',
      search: '',
    };

    const { result } = renderHook(() => useReducer(appReducer, environment));
    const [prevEnv, dispatchEnv] = result.current;
    const { q: prevQ } = prevEnv;

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment: prevEnv, dispatchEnv }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <Sidebar />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );

    const link = wrapper.find('#sidebar-favorites-link').first();
    act(() => {
      link.simulate('click');
    });

    const [env] = result.current;
    const { q } = env;

    expect(q).not.toEqual(!prevQ);
    expect(q.length).toEqual(0);
  });

  it('should have correct hrefs for menu links', () => {
    const mockLocation = {
      pathname: '/favorites',
      search: '',
    };

    const { result } = renderHook(() => useReducer(appReducer, environment));
    const [prevEnv, dispatchEnv] = result.current;

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment: prevEnv, dispatchEnv }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <Sidebar />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );

    const homeLink = wrapper.find('#sidebar-home-link').first();
    const favoritesLink = wrapper.find('#sidebar-favorites-link').first();

    expect(homeLink.prop('to')).toEqual('/');
    expect(favoritesLink.prop('to')).toEqual('/favorites');
  });
});
