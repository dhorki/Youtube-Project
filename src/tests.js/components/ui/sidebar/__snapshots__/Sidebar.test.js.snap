// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar should render correctly with favorites route 1`] = `
<nav
  className="sc-bczRLJ dKvaUX"
>
  <a
    aria-current={null}
    className="sidebar-link "
    href="/"
    id="sidebar-home-link"
    onClick={[Function]}
  >
    <svg
      className="sidebar-link-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
        key="0"
      />
    </svg>
    <span>
      Home
    </span>
  </a>
  <a
    aria-current="page"
    className="sidebar-link sidebar-link-active active"
    href="/favorites"
    id="sidebar-favorites-link"
    onClick={[Function]}
    style={Object {}}
  >
    <svg
      className="sidebar-link-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
        key="0"
      />
      <path
        d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        key="1"
      />
    </svg>
    <span>
      Favorites
    </span>
  </a>
</nav>
`;

exports[`SideBar should render correctly with home route 1`] = `
<nav
  className="sc-bczRLJ dKvaUX"
>
  <a
    aria-current="page"
    className="sidebar-link sidebar-link-active active"
    href="/"
    id="sidebar-home-link"
    onClick={[Function]}
    style={Object {}}
  >
    <svg
      className="sidebar-link-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
        key="0"
      />
    </svg>
    <span>
      Home
    </span>
  </a>
  <a
    aria-current={null}
    className="sidebar-link "
    href="/favorites"
    id="sidebar-favorites-link"
    onClick={[Function]}
  >
    <svg
      className="sidebar-link-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
        key="0"
      />
      <path
        d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        key="1"
      />
    </svg>
    <span>
      Favorites
    </span>
  </a>
</nav>
`;
